turbopack/crates/turbo-persistence/src/db.rs

```diff
@@ -133,156 +133,115 @@ struct Inner {
+        parallel_scheduler: &S,
+    ) -> Result<Self> {
+        match fs::read_dir(path) {
             Ok(entries) => {
-                if !self
-                    .load_directory(entries, read_only)
-                    .context("Loading persistence directory failed")?
+                if let Some(result) =
+                    Self::load_directory(path, entries, read_only, parallel_scheduler)
+                        .with_context(|| {
+                            format!("Loading persistence directory {} failed", path.display())
+                        })?
                 {
-                    if read_only {
-                        bail!("Failed to open database");
-                    }
-                    self.init_directory()
-                        .context("Initializing persistence directory failed")?;
+                    return Ok(result);
                 }
-                Ok(())
+                if read_only {
+                    bail!("Failed to open database {}", path.display());
+                }
+                Self::init_directory(path).with_context(|| {
+                    format!(
+                        "Initializing persistence directory {} failed",
+                        path.display()
+                    )
+                })
             }
             Err(e) => {
                 if !read_only && e.kind() == std::io::ErrorKind::NotFound {
-                    self.create_and_init_directory()
-                        .context("Creating and initializing persistence directory failed")?;
-                    Ok(())
+                    Self::create_and_init_directory(path).with_context(|| {
+                        format!(
+                            "Creating and initializing persistence directory {} failed",
+                            path.display()
+                        )
+                    })
                 } else {
-                    Err(e).context("Failed to open database")
+                    Err(e).with_context(|| {
+                        format!("Reading persistence directory {} failed", path.display())
+                    })
                 }
             }
         }
     }
 
     /// Creates the directory and initializes it.
-    fn create_and_init_directory(&mut self) -> Result<()> {
-        fs::create_dir_all(&self.path)?;
-        self.init_directory()
+    fn create_and_init_directory(path: &Path) -> Result<Self> {
+        fs::create_dir_all(path)?;
+        Self::init_directory(path)
     }
 
     /// Initializes the directory by creating the CURRENT file.
-    fn init_directory(&mut self) -> Result<()> {
-        let mut current = File::create(self.path.join("CURRENT"))?;
+    fn init_directory(path: &Path) -> Result<Self> {
+        let mut current = File::create(path.join("CURRENT"))?;
+        current.try_lock().context(
+            "Failed gain exclusive access to the database. Another process might be using the \
+             database",
+        )?;
```

Comment:
FYI in https://github.com/vercel/next.js/pull/84428 I ended up simulating an advisory lock on Windows using file share modes. Otherwise you get a mandatory lock, which has benefits, but could cause problems for other tools.
https://github.com/vercel/next.js/blob/4bf134433344c57df0ad23a0e7bc81b0a930e990/crates/napi/src/lockfile.rs#L31-L60